ghcide/src/Development/IDE/Core/Compile.hs

```diff
@@ -917,63 +916,16 @@ indexHieFile se mod_summary srcPath !hash hf = do
-    pre style = do
-      tok <- modifyVar indexProgressToken $ fmap dupe . \case
-        x@(Just _) -> pure x
-        -- Create a token if we don't already have one
-        Nothing -> do
-          case lspEnv se of
-            Nothing -> pure Nothing
-            Just env -> LSP.runLspT env $ do
-              u <- LSP.ProgressToken . LSP.InR . T.pack . show . hashUnique <$> liftIO Unique.newUnique
-              -- TODO: Wait for the progress create response to use the token
-              _ <- LSP.sendRequest LSP.SMethod_WindowWorkDoneProgressCreate (LSP.WorkDoneProgressCreateParams u) (const $ pure ())
-              LSP.sendNotification LSP.SMethod_Progress $ LSP.ProgressParams u $
-                toJSON $ LSP.WorkDoneProgressBegin
-                  { _kind = LSP.AString @"begin"
-                  ,  _title = "Indexing"
-                  , _cancellable = Nothing
-                  , _message = Nothing
-                  , _percentage = Nothing
-                  }
-              pure (Just u)
-
-      (!done, !remaining) <- atomically $ do
-        done <- readTVar indexCompleted
-        remaining <- HashMap.size <$> readTVar indexPending
-        pure (done, remaining)
-      let
-        progressFrac :: Double
-        progressFrac = fromIntegral done / fromIntegral (done + remaining)
-        progressPct :: LSP.UInt
-        progressPct = floor $ 100 * progressFrac
-
-      whenJust (lspEnv se) $ \env -> whenJust tok $ \token -> LSP.runLspT env $
-        LSP.sendNotification LSP.SMethod_Progress $ LSP.ProgressParams token $
-          toJSON $
-            case style of
-                Percentage -> LSP.WorkDoneProgressReport
-                    { _kind = LSP.AString @"report"
-                    , _cancellable = Nothing
-                    , _message = Nothing
-                    , _percentage = Just progressPct
-                    }
-                Explicit -> LSP.WorkDoneProgressReport
-                    { _kind = LSP.AString @"report"
-                    , _cancellable = Nothing
-                    , _message = Just $
-                        T.pack " (" <> T.pack (show done) <> "/" <> T.pack (show $ done + remaining) <> ")..."
-                    , _percentage = Nothing
-                    }
-                NoProgress -> LSP.WorkDoneProgressReport
-                  { _kind = LSP.AString @"report"
-                  , _cancellable = Nothing
-                  , _message = Nothing
-                  , _percentage = Nothing
-                  }
-
+    pre optProgressStyle = modifyVar_ indexProgressToken $ fmap Just . \case
+              Just x -> return x
+              -- create a progressReport if we don't already have one
+              Nothing -> do
+                  tt <- progressReportingOutsideState
+                    (liftM2 (+) (HashMap.size <$> readTVar indexPending) (readTVar indexCompleted))
+                    (readTVar indexCompleted)
```

Comment:
This seems a bit backwards to me? Can we not use the usual `ProgressReporting` and then call `progressUpdate` etc. instead of touching these TVars in the indexing code?

Reply:
Okay, I think I get it. Now I wonder if it would be simpler to make the other progress tracker work this way?

Reply:
As along as they are already tracking the progress status external. 

But I take a look at the use case at `kick`, the state(todo count,done count) tracking is at `defineEarlyCutoff'` and independent for each file. No easy way to swtich the `Kick` progress tracker work this way.

Reply:
Right, but when we _create_ the progress tracker we could set that stuff up externally, so that the `ProgressReporting` doesn't see it. Let me see if I can just write it down.

Reply:
Here: https://github.com/haskell/haskell-language-server/commit/2ad92a01e89480e06d60647f127fae052054c71e

If we did this then I think `indexHieFile` could just use `progressCounter` directly?

Reply:
Oh I see. I've done that to the `loop`, but exract that to `progressCounter` is cleaner

Reply:
Yeah, it's a nice standalone thing in fact

Reply:
I ported 2ad92a0 here

Reply:
then I think if you export it you can just use it in `indexHieFile`

Reply:
Great, extract it to standalone now! @michaelpj

Reply:
If using it barely in `indexHieFile`, we still need some kind of wrapping around `progressCounter`.  `async` and `cancel` the `progressCounter`.

Reply:
Yeah, we do need that. But we could just do it locally. Or maybe you want to have a standalone concept of a "resumable/cancellable progress counter" and then just put the file-tracking on top for the compilation progress tracker?

Reply:
May be the latter, I am moving up the `indexHieFile` reporting initilization to `shakeOpen` as that of  the `Kick` reporting. Should be much cleaner than the preivous version.

Reply:
"resumable/cancellable progress counter" may need the additional Event `ProgressTryToStart`.